openstax/mathmlcloud
# Upgrade node and packages

I spend quite a bit of time separating changes into commits.  Please :bowing_woman: make sure github **doesn't squash** this PR into one commit!

---

- Upgrade base docker image to node 11 (latest)


- Update packages to latest where possible

  Some of the packages are no longer maintained.  Will need to replace
  them later.

- Import ObjectId from mongodb in FeedbackController.js

  'sails-mongo/node_modules/mongodb' is not found.  After looking at the
  code in sails-mongo, the correct way of importing ObjectId is:
  
  ```
  ObjectId = require('mongodb').ObjectID || require('mongodb').ObjectId;
  ```

- Upgrade packages and update code to use sails 1.0

  Some of the code changes were done by the sails automated migration tool
  and some were manual changes by reading the migration guide:
  
  https://sailsjs.com/documentation/upgrading/to-v-1-0
  
  Changes include:
  
   - `.add`, `.create`, `.save` on model instances no longer exist and has
     been replaced by model static methods.
   - `connections` has been renamed to `datastores`.
   - In User model, `defaultsTo` has been removed because `defaultsTo` and
     `required: true` are mutually exclusive.
   - In User model, `organizationTypes` has been changed to
     `type: 'json'`, `columnType: 'array'` because `type: array` doesn't
     exist anymore
   - User model custom `toJSON` becomes `customToJSON`.
   - Use `middleware` instead of `customMiddleware` in config/passport.js
     and add `passportInit` to config/http.js.

- Add html to conversion option for POST /equation

  It was only possible to convert to "svg", "png" or "mml" using
  `POST /equation`.  Add "html" to the API so it returns the html and css
  for the equation.

- Fix User model .customToJSON to use _.omit

  Building a hash from the user model using `.toObject()` is causing
  some unknown errors and so changing to use `_.omit` instead to remove
  password hash from the user json.

- Pass in pageNum and pageSize to .paginate

  When accessing the feedback API, sails outputs this in the console:
  
  Deprecation warning: Passing in a dictionary (plain JS object) to .paginate()
  is no longer supported -- instead, please use:
  ```
  .paginate(pageNum, pageSize)
  ```

- Fix feedback API to use query with explicit "where"

  Whenever the query contains "limit", "skip" or "sort", we need to use a
  "where" key to specific attributes. See:
  
  https://sailsjs.com/documentation/concepts/models-and-orm/query-language

- Change all ".jsonx" to ".json" in api/responses

  `.jsonx` is deprecated in sails 1.0 and will be removed in future
  releases.  According to the upgrade doc
  https://sailsjs.com/documentation/upgrading/to-v-1-0
  we can just replace `res.jsonx()` in `api/responses` with `res.json()`.

- Use mathjax-node-sre so speakText is available

  mathjax-node removed their integration with speech-rule-engine:
  
  https://github.com/mathjax/MathJax-node/issues/207
  
  This means that we no longer get speakText support for equations, like
  "a plus b" for "a + b".
  
  mathjax-node-sre is a wrapper around mathjax-node and can be used as a
  drop-in replacement for mathjax-node.
- Change default page to 0 where .paginate is used

  The default is 1 which doesn't return any results if the number of
  results is less than the page size.  `.paginate` is probably expecting
  0 instead of 1.

- Change Feedback.submittedBy from req.user to req.user.id

  When submitting a feedback for an equation, if the `submittedBy` field
  is `req.user`, the feedback does not appear in the submitter's
  `/myFeedback` page.  The `submittedBy` field should use the user id
  instead.

- Add cucumber and gherkin as dev dependencies

  Cucumber tests are in test/features.  `cucumber` and `gherkin` need to
  be installed to run these tests.

- Update tests in spec/ to work with frisby 2.1.0

  Changes include:
  - Changing `frisby.create()` to `it()`
  - Changing `frisby.create().post()` and `.get()` to `frisby.post()` and
    `frisby.get()`
  - Adding `role: 'user'` to all user creation requests
  - Changing `.afterJSON()` to `.then()`
  - Changing `.expectStatus()`, `.expectJSON()`, `.expectHeaderContains()`
    etc to `.expect()`
  - Changing `.toss()` to `.done()`
  - Changing expected content types to include `charset=utf-8`
  - Disabling unimplemented tests with `xit`

- Change cucumber tests to use localhost:1337

  Change cucumber feature tests from using "http://192.168.99.100:1337" to
  "http://localhost:1337".

- Update feature tests to work with cucumber 5.0.3

  Changes include:
  - using `Given()`, `When()` and `Then()` from cucumber
  - using `setWorldConstructor()` from cucumber to set up a "world" -- an
    isolated context for each scenario, exposed to steps as `this`
  - implementing TODOs

- Implement error response feature test

  The error response feature tests was written in gherkin but did not have
  a step definition file with code to support it.

- Update code for jsdom 13.0.0 in ConversionService.js

  The API has changed and now we need to use `new JSDOM()` and
  `.serialize()`.

- Shutdown jobs queue and stop kue app server on sails.lower

  We use "kue" to create a queue for html5 page conversions.  There is a
  server listening on port 3000 in bootstrap.js when sails starts up, but
  there is no code to shutdown the server when sails is shutdown.
  
  This adds a hook on `sails.lower` to shutdown the jobs queue and the kue
  app server.

- Add "test" to package.json "scripts"

  This allows us to run the tests by running `npm test`.  There are two
  types of tests.  First we run the specs using jasmine, then we run the
  feature tests using cucumber.  The cucumber tests are only run if
  jasmine tests pass.

- Add code to start sails server in tests

  Jasmine and cucumber tests expect a server to be running.  It's best to
  include the code to start and stop the server in the test set up code so
  it is easy to run.  For jasmine tests, we add event handlers for
  jasmineStarted and jasmineDone events.  For cucumber tests, we add
  `BeforeAll` and `AfterAll` handlers.

- Add clean up code to remove users created in tests

  Tests create users with random usernames.  To avoid the build up of test
  data, we are removing the users created in tests after the tests.

- Add travis config file to run npm test

  Since an important part of this project is the docker image, we are
  going to build the docker image and run the tests inside the docker
  container.

- Remove tests and spec from .dockerignore

  We are running tests in the docker container on travis, so really need
  to copy the test and spec directories to the container.  It should not
  cause any problems having them in the container.

---

I spend quite a bit of time separating changes into commits.  Please :bowing_woman: make sure github **doesn't squash** this PR into one commit!
